kubernetes/metrics.go

```diff
@@ -0,0 +1,264 @@
+	return clientset, err
+}
+
+// GetRawMetrics returns list of metrics for a given namespace
+func GetRawMetrics(configPath string, namespace string, pods *PodMetricsList) error {
+	clientset, err := K8sClient.Get(configPath)
+	if err != nil {
+		return err
+	}
+	data, err := clientset.RESTClient().Get().AbsPath("apis/metrics.k8s.io/v1beta1/namespaces/" + namespace + "/pods").DoRaw()
+	if err != nil {
+		return err
+	}
+	logging.Info("----raw json: %v", string(data))
+	err = json.Unmarshal(data, &pods)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetProcessedMetrics extracts metrics from k8s using GetRawMetrics and then populates them with labels and
+// float metrics that are converted from raw metrics' string values
+func GetProcessedMetrics(configPath string, namespace string, pods *PodMetricsList) error {
+	if err := GetRawMetrics(configPath, namespace, pods); err != nil {
+		return err
+	}
+
+	if err := GetLabels(configPath, namespace, pods); err != nil {
+		return err
+	}
+
+	ProcessMetrics(&pods)
+
+	return nil
+}
+
+// GetAverageMetrics extract metrics from k8s using GetRawMetrics and then transforms them to
+// new structure with labels and average CPU and memory per pod
+func GetAverageMetrics(configPath string, namespace string) ([]PodAverageMetrics, error) {
+	var pods PodMetricsList
+
+	if err := GetRawMetrics(configPath, namespace, &pods); err != nil {
+		return nil, err
+	}
+
+	if err := GetLabels(configPath, namespace, &pods); err != nil {
+		return nil, err
+	}
+
+	return CalculateAverageMetrics(&pods)
+}
+
+// GetLabels populates pod metrics list with labels
+func GetLabels(configPath string, namespace string, pods *PodMetricsList) error {
+	clientset, err := K8sClient.Get(configPath)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(pods.Items); i++ {
+		if len(pods.Items[i].Metadata.Labels) == 0 {
+			logging.Info("----getting labels for %v", pods.Items[i].Metadata.Name)
+			pod, err := clientset.CoreV1().Pods(namespace).Get(pods.Items[i].Metadata.Name, metav1.GetOptions{})
+			logging.Info("----got pod data: %v", pod)
+			if err != nil {
+				return err
+			}
+			pods.Items[i].Metadata.Labels = pod.Labels
+		}
+	}
+
+	return nil
+}
+
+// PodAverageMetrics provides average CPU and memor as long as some pod's metadata
+type PodAverageMetrics struct {
+	Name   string
+	Labels map[string]string
+	CPU    float64
+	Memory float64
+}
+
+func CalculateAverageMetrics(pods *PodMetricsList) ([]PodAverageMetrics, error) {
+	var res = make([]PodAverageMetrics, len(pods.Items))
+	for i := 0; i < len(pods.Items); i++ {
+		res[i].Name = pods.Items[i].Metadata.Name
+		res[i].Labels = pods.Items[i].Metadata.Labels
+		var sumCPU, sumMem float64
+		for j := 0; j < len(pods.Items[i].Containers); j++ {
+			cpu, err := ConvertCPU(pods.Items[i].Containers[j].Usage.CPU)
+			if err == nil {
+				sumCPU += cpu
+			} else {
+				logging.Error("Conversion of CPU for %v failed - %v", pods.Items[i], err)
+				return nil, err
```

Comment:
I think returning error on one conversion can be too much, on startups of pods and similar events, there is usually empty or 0 values.

Reply:
Fixed